evm/src/executor/inspector/cheatcodes/expect.rs

```diff
@@ -99,6 +141,77 @@ pub fn handle_expect_revert(
+) -> String {
+    match expected_revert.captured_revert_data {
+        Some(captured_revert_data) => {
+            if captured_revert_data.is_empty() {
+                format!(
+                    "The expected revert address {:#?} reverted as expected, but without data",
+                    expected_revert.address
+                )
+            } else if let Some(expected_revert_reason) = expected_revert.reason {
+                format!(
+                "The expected revert address {:#?} did not revert with the expected revert data. Expected: {} but found: {}",
+                expected_revert.address,
+                stringify(&expected_revert_reason),
+                stringify(&captured_revert_data),
+            )
+            }
+            // Technically this scenario shouldn't happen because if the expected revert data is
+            // None and the address of the revert matches then the test should pass.
+            else {
+                format!(
+                "The expected revert address {:#?} did not revert with the expected revert data.",
+                expected_revert.address
+            )
+            }
+        }
+        None => {
+            format!("The expected revert address {:#?} did not revert", expected_revert.address,)
+        }
+    }
+}
+
+/// Verifies that a revert matches one of the reverts associated to an address as expected by the
+/// user and that it happened in the expected order
+pub fn handle_expected_reverts_with_address(
+    state: &mut Cheatcodes,
+    current_contract: H160,
+    current_revert_data: Bytes,
+) {
+    // If this is empty we don't need to match any revert
+    if state.expected_reverts_with_address.is_empty() {
+        return
+    }
+
+    // Take the first expected revert and...
+    if let Some(mut first) = state.expected_reverts_with_address.pop_front() {
+        // ... verify if it has been alredy matched.
+        if !first.found {
+            // if it matches with the current revert mark found as true and push the current
+            // expected revert to the end of the queue to match any other pending expected
+            // revert.
+            if first.compare_revert(current_contract, current_revert_data) {
+                first.found = true;
+                state.expected_reverts_with_address.push_back(first);
+            }
+            // If the current expected revert does not match we push it to the front of the queue to
+            // try to match it again with the next revert.
+            else {
+                state.expected_reverts_with_address.push_front(first);
+            }
+        }
+        // If the first expected revert has already been matched it means that all the expected
+        // reverts have been matched and we can safely clear the queque.
+        else {
+            state.matched_all_expected_reverts_with_address = true;
+            state.expected_reverts_with_address.clear();
+        }
```

Comment:
I still don't really understand this—we're only really matching one event? So why not use an `Option` here? The vecdequeue can only hold at most one event. because we're clearing it here.

Reply:
We are matching multiple reverts because the user can call `expectRevert` with an address multiple times. So every time a revert happens it is compared against the first revert in the deque (because order matters).
If the revert matches with the first one then it is moved at the end of the queue, if not it is put in front once again until a matching revert is found. Following this cycle until the end of the call we could have the following scenarios:
- The first element of the deque has been matched which means that all the other reverts after it have been matched too. This is because at some point all the reverts that have been matched have been moved at the end of the deque and the first one that initially wasn't matched is brought back to the front.
- The first element of the dequeue has still not been matched which means that the expected revert was not thrown during EVM execution.